sklearn/semi_supervised/_self_training.py

```diff
@@ -0,0 +1,326 @@
+import numpy as np
+
+from ..base import MetaEstimatorMixin, clone
+from ..utils.validation import check_X_y, check_array, check_is_fitted
+from ..utils.metaestimators import if_delegate_has_method
+from ..utils import safe_mask
+from ..exceptions import ConvergenceWarning
+from ..utils.metaestimators import _BaseComposition
+import warnings
+
+__all__ = ["SelfTrainingClassifier"]
+
+# Authors: Oliver Rausch   <rauscho@ethz.ch>
+#          Patrice Becker  <beckerp@ethz.ch>
+# License: BSD 3 clause
+
+
+def _validate_estimator(estimator):
+    """Make sure that an estimator implements the necessary methods."""
+    if not hasattr(estimator, "predict_proba"):
+        msg = "base_classifier ({}) should implement predict_proba!"
+        raise ValueError(msg.format(type(estimator).__name__))
+
+
+class SelfTrainingClassifier(MetaEstimatorMixin, _BaseComposition):
+    """Self-training classifier
+
+    This class allows a given supervised classifier to function as a
+    semi-supervised classifier, allowing it to learn from unlabeled data. It
+    does this by iteratively predicting labels for the unlabeled data.
+
+    Read more in the :ref:`User Guide <self_training>`.
+
+    Parameters
+    ----------
+    base_classifier : estimator object
+        An estimator object implementing ``fit`` and ``predict_proba``.
+        Invoking the ``fit`` method will fit a clone of the passed estimator,
+        which will be stored in the ``base_classifier_`` attribute.
+
+    threshold : float, optional (default=0.75)
+        The decision threshold. If the ``base_classifier`` makes a prediction
+        with a ``predict_proba`` above this threshold, it will be added to the
+        labeled dataset.
+        Should be in [0, 1).
+
+    max_iter : int or ``None``, optional (default=10)
+        Maximum number of iterations allowed. Should be greater than or equal
+        to 0. If it is ``None``, the classifier will continue to predict labels
+        until all unlabeled samples have been labeled. In this case, be aware
+        that the fit may never terminate.
+
+    n_iter_no_change: int, optional (default=1)
+        If not `None`, early stopping will be applied. In this case, the
+        classifier will count the amount of new labels in each iteration. If
+        no new labels are added to the training set for `n_iter_no_change`
+        iterations, the classifier will stop fitting early.
+
+    verbose: bool, (default=False)
+        Enable verbose output.
+
+    Attributes
+    ----------
+    base_classifier_ : estimator object
+        The fitted estimator.
+
+    y_labels_ : array, shape=(n_samples,)
```

Comment:
LabelPropagation calls this `transduction_`. At least, I don't like the `y_` prefix.